libvast/src/directory.cpp

```diff
@@ -128,32 +130,31 @@ caf::expected<size_t> recursive_size(const std::filesystem::path& root_dir) {
   return total_size;
 }
 
-std::vector<path>
-filter_dir(const path& dir, std::function<bool(const path&)> filter,
+caf::expected<std::vector<std::filesystem::path>>
+filter_dir(const std::filesystem::path& root_dir,
+           std::function<bool(const std::filesystem::path&)> filter,
            size_t max_recursion) {
-  std::vector<path> result;
-  if (max_recursion == 0)
-    return result;
-  for (auto& f : directory(dir)) {
-    switch (f.kind()) {
-      default: {
-        if (!filter || filter(f))
-          result.push_back(f);
-        break;
-      }
-      case path::directory: {
-        // Recurse directories depth-first.
-        auto paths = filter_dir(f, filter, --max_recursion);
-        if (!paths.empty()) {
-          auto begin = std::make_move_iterator(paths.begin());
-          auto end = std::make_move_iterator(paths.end());
-          result.insert(result.end(), begin, end);
-          std::sort(result.begin(), result.end());
-        }
-        break;
-      }
-    }
+  std::vector<std::filesystem::path> result;
+  std::error_code ec{};
+  auto dir = std::filesystem::recursive_directory_iterator(root_dir, ec);
+  if (ec)
+    return caf::make_error(ec::filesystem_error, ec.message());
+  auto begin = std::filesystem::begin(dir);
+  auto end = std::filesystem::end(dir);
+  result.reserve(std::distance(begin, end));
+  while (begin != end) {
+    const auto current_path = begin->path();
+    const auto current_depth = static_cast<size_t>(begin.depth());
+    if (current_depth >= max_recursion)
+      return caf::make_error(ec::recursion_limit_reached,
+                             fmt::format("reached recursion limit when "
+                                         "filtering directory {}",
+                                         root_dir.string()));
```

Comment:
```suggestion
                                         root_dir));
```

I fixed this in #1416 today.

Reply:
Thanks for the formatting fix. I saw the errors yesterday, but didn't dig into why, which is why I just kept it simple and got the underlying string representation for the `std::filesystem::path`. I prefer sticking with strong types, so good fix 👍

Reply:
I went ahead and fixed the usage here and a few other places to not call the `string()` member function on the `path` when formatting it.

Reply:
Yeah, that fix was rather complicated, but I think we should be all good now.

We have an ongoing effort currently to move integrate {fmt} more deeply as it's a rather new dependency, so there's a lot happening on that front currently. We also intend to poison `caf::make_error` and provide a custom `make_error` function that always goes through `fmt::format` to ensure consistent formatting.

Thanks for your contribution—again! Will merge in a bit. ❤️

Reply:
@JoeLoser I had to push another commit to fix a segfault. Maybe you know why this happens, but at least I managed to fix it.

Reply:
Hmm, I'm not sure why it happens. [Here is a simple Wandbox](https://wandbox.org/permlink/WcDsrIrV9uIsENT7) showing `std::distance` working with `std::filesystem::recursive_directory_iterator`.

At first, when I read your comment, perhaps it is due to how `std::filesystem::end` returns a default constructed std::filesystem::recursive_directory_iterator` -- a bit strange as it ignores its argument, which is different from canonical `end` function works for most types. I was thinking that wouldn't play nicely with `std::advance()`, but the Wandbox link above shows it should work fine I think. 🤔 

The integration tests look to be failing still, but I haven't yet had luck running them locally on Mac for some reason that I haven't dug into yet -- I've been doing development just with the unit tests locally so far.

Reply:
I see why it segfaults. Note that `std::filesystem::recursive_directory_iterator` and `std::filesystem::directory_iterator` are both input iterators. This means that after you call `std::distance`, you cannot use the begin iterator again. So the while loop is UB.

So removing the `std::distance` call for reserving the size is the right play. As to why the integration tests are still failing after your commit, I'm not sure.

Reply:
Makes sense.

The integration tests are failing because filter_dir must sort its result lexicographically.

Reply:
I meant to ask about that last night -- I was wondering what forced that requirement and whether we needed that behavior in the refactoring.

Reply:
It's needed so the schema files are read in a fixed order. Not actually sure if that's a hard requirement after our recent changes on the type system in VAST's query and schema language, but let's not change that as part of this PR.

Reply:
I see. I just made it so the filtered directory is sorted, so the integration tests should be good now. 

Thanks for the explanation for the use case/requirements.